cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java

```diff
@@ -457,7 +457,8 @@ public KafkaAssemblyOperator(Vertx vertx, boolean isOpenShift,
         Future<ReconciliationState> zkRollingUpdate() {
             return withVoid(zkSetOperations.maybeRollingUpdate(zkDiffs.resource(), pod ->
                 isPodToRestart(zkDiffs.resource(), pod,
-                        zkAncillaryCmChange, this.clusterCa.certRenewed(), this.clusterCa.certsRemoved())
+                        zkAncillaryCmChange, this.clusterCa.certRenewed(), this.clusterCa.certsRemoved(),
+                        this.clientsCa.certRenewed(), this.clientsCa.certsRemoved())
```

Comment:
Why not just pass the whole `clientsCa` and `clusterCa`, and put the logic about certs removed or renewed directly in `isPodToRestart()`?

Reply:
Makes sense.

Reply:
Actually, sort-of does and sort-of doesn't. There should be no need to roll ZK if the clients CA changes, since ZK has no dependency (trust relationship) on the clients CA. So `isPodToRestart()` would need to take a `Ca...` parameter, and here we'd call it just with the `clusterCa`, but for the `kafkaRollingUpdate()` we'd call it with both `clientsCa` and `clusterCa`.

Reply:
My only concern is about the logic that we have to put into the `isPodToRestart` to check how many `Ca`(s) we have and if one of them is a `ClientsCa` and the distinguish the logging part accordingly. Does it worth the effort?

Reply:
I don't understand. `isPodToRestart()` would be passed a `Ca[]` and would simply loop over them to determine whether any had a renewed cert or replace key or whatever. `isPodToRestart()` itself wouldn't need care about what sort of CA it might be causing the restart. It's at the call site whether the different dependencies are expressed as passing either just the cluster CA (for ZK case), or both cluster and clients CA (for Kafka case). What am I missing?

Reply:
It could be something like following ... 

```
private boolean isPodToRestart(StatefulSet ss, Pod pod, boolean isAncillaryCmChange, Ca... cas) {
            boolean isPodUpToDate = isPodUpToDate(ss, pod);
            boolean isPodCaCertUpToDate = isPodCaCertUpToDate(pod);
            boolean isCaCertsChanged = false;

            if (log.isDebugEnabled()) {
                String reason = "";
                for (Ca ca: cas) {
                    if (ca instanceof ClientsCa) {
                        if (ca.certRenewed()) {
                            reason += "clients CA certificate renewal, ";
                        }
                        if (ca.certsRemoved()) {
                            reason += "clients CA certificate removal, ";
                        }
                    } else {
                        if (ca.certRenewed()) {
                            reason += "cluster CA certificate renewal, ";
                        }
                        if (ca.certsRemoved()) {
                            reason += "cluster CA certificate removal, ";
                        }
                    }
                }
                if (isAncillaryCmChange) {
                    reason += "ancillary CM change, ";
                }
                if (!isPodUpToDate) {
                    reason += "Pod has old generation, ";
                }
                if (!isPodCaCertUpToDate) {
                    reason += "Pod has old cluster CA certificate generation";
                }
                if (!reason.isEmpty()) {
                    log.debug("{}: Rolling pod {} due to {}",
                            reconciliation, pod.getMetadata().getName(), reason.substring(0, reason.length() - 2));
                }
            }

            for (Ca ca: cas) {
                isCaCertsChanged = isCaCertsChanged || ca.certRenewed() || ca.certsRemoved();
            }
            

            return !isPodUpToDate || !isPodCaCertUpToDate || isAncillaryCmChange || isCaCertsChanged;
        }
```

or you should assume that `cas[0]` is the needed Cluster CA and the `cas[1]` is the optional Clients CA so checking the length.
But one more time, does it add any real value and it's worth doing that?

Reply:
I didn't notice the `toString` override in order to address this logging "issue" :-)